client/client.go

```diff
@@ -0,0 +1,161 @@
+	for _, opt := range opts {
+		opt(options)
+	}
+
+	inv, err := generateInvocation(methodName, paramsRawVals, callType, options)
+	if err != nil {
+		return nil, err
+	}
+	// todo: move timeout into context or invocation
+	return cli.invoker.Invoke(ctx, inv), nil
+
+}
+
+func (cli *Client) CallUnary(ctx context.Context, req, resp interface{}, interfaceName, methodName string, opts ...CallOption) error {
+	res, err := cli.call(ctx, []interface{}{req, resp}, interfaceName, methodName, constant.CallUnary, opts...)
+	if err != nil {
+		return err
+	}
+	return res.Error()
+}
+
+func (cli *Client) CallClientStream(ctx context.Context, interfaceName, methodName string, opts ...CallOption) (interface{}, error) {
+	res, err := cli.call(ctx, nil, interfaceName, methodName, constant.CallClientStream, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return res.Result(), res.Error()
+}
+
+func (cli *Client) CallServerStream(ctx context.Context, req interface{}, interfaceName, methodName string, opts ...CallOption) (interface{}, error) {
+	res, err := cli.call(ctx, []interface{}{req}, interfaceName, methodName, constant.CallServerStream, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return res.Result(), res.Error()
+}
+
+func (cli *Client) CallBidiStream(ctx context.Context, interfaceName, methodName string, opts ...CallOption) (interface{}, error) {
+	res, err := cli.call(ctx, nil, interfaceName, methodName, constant.CallBidiStream, opts...)
+	if err != nil {
+		return nil, err
+	}
+	return res.Result(), res.Error()
+}
+
+func (cli *Client) Init(info *ClientInfo) error {
+	if info == nil {
+		return errors.New("ClientInfo is nil")
+	}
+	cli.cfg.ReferWithInfo(info)
+	cli.invoker = cli.cfg.invoker
+
+	return nil
+}
+
+func generateInvocation(methodName string, paramsRawVals []interface{}, callType string, opts *CallOptions) (protocol.Invocation, error) {
+	inv := invocation_impl.NewRPCInvocationWithOptions(
+		invocation_impl.WithMethodName(methodName),
+		// todo: process opts
+		invocation_impl.WithParameterRawValues(paramsRawVals),
+	)
+	inv.SetAttribute(constant.CallTypeKey, callType)
+
+	return inv, nil
+}
+
+func NewClient(opts ...ReferenceOption) (*Client, error) {
+	// get default RootConfigs
+	rootCfg := config.NewRootConfigBuilder().Build()
+	rootCfg.Init()
+	newRefCfg := &ReferenceConfig{}
+	for _, opt := range opts {
+		opt(newRefCfg)
+	}
+	if err := newRefCfg.Init(rootCfg); err != nil {
+		return nil, err
+	}
+	return &Client{
+		cfg: newRefCfg,
+	}, nil
+}
+
+// NewClientWithReferenceBase is used by /config/consumer_config, it assumes that refCfg has been init
+func NewClientWithReferenceBase(refCfg *ReferenceConfig, opts ...ReferenceOption) (*Client, error) {
+	if refCfg == nil {
+		return nil, errors.New("ReferenceConfig is nil")
+	}
+	for _, opt := range opts {
+		opt(refCfg)
+	}
+	return &Client{
+		cfg: refCfg,
+	}, nil
+}
+
+func NewConsumerConfig(opts ...ConsumerOption) (*ConsumerConfig, error) {
```

Comment:
I think it's better to expose only `NewClientXxx` varies in this package to users. Move `NewConsumerConfig` and `NewMethodConfig` to a new place or the previous place where it is from.